app/js/lib/foxbox/services.js

```diff
@@ -0,0 +1,273 @@
+      this[p.nextPollTimeout] = null;
+    }
+  }
+
+  /**
+   * Returns in-memory service cache. When cache is not initialized we fill it
+   * in with the content of local IndexedDB.
+   *
+   * @return {Promise<Map<string, BaseService>>}
+   * @private
+   */
+  [p.getCache]() {
+    if (this[p.cache]) {
+      return Promise.resolve(this[p.cache]);
+    }
+
+    const cache = this[p.cache] = new Map();
+
+    return this[p.db].getServices()
+      .then((dbServices) => {
+        dbServices.forEach(({ data: dbService }) => {
+          cache.set(dbService.id, this[p.getServiceInstance](dbService));
+        });
+
+        return cache;
+      });
+  }
+
+  /**
+   * Schedules an attempt to poll the server, does nothing if polling is not
+   * enabled or it has already been scheduled. New poll is scheduled only once
+   * previous one is completed or failed.
+   *
+   * @private
+   */
+  [p.schedulePoll]() {
+    // Return early if polling is not enabled or it has already been scheduled.
+    if (!this[p.isPollingEnabled] || this[p.nextPollTimeout]) {
+      return;
+    }
+
+    this[p.nextPollTimeout] = setTimeout(() => {
+      Promise.all([this[p.db].getServices(), this[p.api].get('services')])
+        .then(([storedServices, fetchedServices]) => {
+          return this[p.updateServiceList](storedServices, fetchedServices);
+        })
+        .catch((error) => {
+          console.error(
+            'Polling has failed, scheduling one more attempt: ',
+            error
+          );
+        })
+        .then(() => {
+          this[p.nextPollTimeout] = null;
+          this[p.schedulePoll]();
+        });
+    }, this[p.settings].pollingInterval);
+  }
+
+  /**
+   * Tries to update service list and emits appropriate events.
+   *
+   * @param {Array<Object>} storedServices Services currently stored in DB.
+   * @param {Array<Object>} fetchedServices Services returned from the server.
+   * @return {Promise}
+   * @private
+   */
+  [p.updateServiceList](storedServices, fetchedServices) {
+    return this[p.getCache]()
+      .then((cache) => {
+        let servicesToAddCount = 0;
+        fetchedServices.forEach((fetchedService) => {
+          const storedService = storedServices.find(
+            (storedService) => storedService.data.id === fetchedService.id
+          );
+
+          const isExistingService = !!storedService;
+
+          if (isExistingService &&
+              isSimilar(fetchedService, storedService.data)) {
+            return;
+          }
+
+          // Populate the db with the latest service.
+          this[p.db].setService(fetchedService);
+
+          const service = this[p.getServiceInstance](fetchedService);
+          cache.set(service.id, service);
+
+          if (isExistingService) {
+            this.emit('service-changed', service);
+          } else {
+            servicesToAddCount++;
+          }
+        });
+
+        const servicesToRemoveCount = storedServices.length +
```

Comment:
A bit nasty calculation, but it allows us to avoid unnecessary iterations over stored services, so we only do it when we really need to remove anything.